mirrord-cli/src/main.rs

```diff
@@ -64,6 +64,83 @@ const INJECTION_ENV_VAR: &str = "LD_PRELOAD";
 #[cfg(target_os = "macos")]
 const INJECTION_ENV_VAR: &str = "DYLD_INSERT_LIBRARIES";
 
+/// This is for Apple M1 running mirrord aarch trying to execute x86 binaries.
+#[cfg(all(target_os = "macos", target_arch = "aarch64"))]
+mod mac_aarch {
+    use std::{
+        io::{Cursor, Read},
+        path::Path,
+    };
+
+    use mach_object::{OFile, CPU_TYPE_X86_64};
+    use reqwest;
+    use search_path::SearchPath;
+
+    use super::*;
+    pub fn is_binary_different_arch(binary_path: &String) -> bool {
+        let search_path = SearchPath::new("PATH").unwrap();
+        let binary_path = match search_path.find(&Path::new(binary_path)) {
+            Some(path) => path,
+            None => {
+                warn!("Could not find binary in PATH");
+                return false;
+            }
+        };
+        let mut binary_file = match File::open(binary_path) {
+            Ok(file) => file,
+            Err(err) => {
+                warn!("Could not open binary: {err:?}");
+                return false;
+            }
+        };
+        let mut binary_content = Vec::new();
+        let size = match binary_file.read_to_end(&mut binary_content) {
+            Ok(size) => size,
+            Err(err) => {
+                warn!("Could not read binary: {err:?}");
+                return false;
+            }
+        };
+        let mut binary_cursor = Cursor::new(&binary_content[..size]);
+        match OFile::parse(&mut binary_cursor) {
+            Ok(OFile::MachFile {
+                header,
+                commands: _,
+            }) => header.cputype == CPU_TYPE_X86_64,
+            Ok(_) => {
+                warn!("Could not parse binary as Mach-O File");
+                false
+            }
+            Err(err) => {
+                warn!("Could not open binary: {err:?}");
+                false
+            }
+        }
+    }
+
+    pub fn extract_intel_binary() -> Result<String> {
```

Comment:
Rename it to `extract_or_create_intel_binary` to make it a bit clearer what might happen here.